### DISCUSSION TITLE
OceanBase Community Monthly Report 2022/4

### DISCUSSION THREAD

**Discussion by longdafeng** [oceanbase/oceanbase#871] [Announcements]
Reserver for Summary

**Comment by longdafeng** [comment]
# 2022 RoadMap 大纲
OceanBase 社区版发布节奏如下：
每年一个大版本 release，每3个月一次 feature 版本, 每个月一个 bug fix 版本( bp 版本).
1. 大版本发布即为架构发生升级， 版本升级类似 MySQL 5.7 升级到 MySQL 8.0, 需要做数据迁移操作
2. feature 版本即为发布了众多 feature 或大 feature , 本地手动冷升级(本地重启)或者通过 OCP 热升级(不停服务)
3. bp 版本即为纯 bug fix 版本, 版本升级直接替换 binary 即可, 可以使用 ODP 直接 binary 替换或使用 OCP 热升级


# Roadmap

![image](https://user-images.githubusercontent.com/374091/163750747-ad0ca54b-047a-4fff-b15f-0ba8d944e54f.png)

**Comment by longdafeng** [comment]
# 研发进度
## 内核研发

| **需求来源** | **                 目标和任务详细描述** | **进展详细描述** |
| --- | --- | --- |
| 六棱镜 | [数据文件按需申请](https://github.com/oceanbase/oceanbase/issues/875) | 编码中 |
| 钉钉用户 | [建议添加list分区根据范围查询可以使用到分区裁剪的功能](https://github.com/oceanbase/oceanbase/issues/840) | 编码中 |
| 红象云腾 | [support latin1 character set](https://github.com/oceanbase/oceanbase/issues/689) | 详细设计中 |
| 携程  | [审计日志gv$sql_audit 数据持久化](https://github.com/oceanbase/oceanbase/issues/873) | 概要设计中 |
| 用友 | [打开PS, 更好支持golang](https://github.com/oceanbase/oceanbase/issues/331) | 概要设计中 |
| 京东 | [DML执行失败时, 快速定位失败的列](https://github.com/oceanbase/oceanbase/issues/817) | 概要设计中 |
| 基石数据 | [监控数据增加直方图类型, 更便捷输出至Prometheus](https://github.com/oceanbase/oceanbase/issues/767) | 概要设计中 |
| 贝壳 | [多模数据库（Table、HBase API）需要支持自动过期数据清理](https://github.com/oceanbase/oceanbase/issues/874) | 概要设计中 |
| 贵州双龙 | [data_dir参数值不准确](https://github.com/oceanbase/oceanbase/issues/876) | 概要设计中 |
| 快手 | [用户看到的错误码和日志中的一致](https://github.com/oceanbase/oceanbase/issues/877) | 概要设计中 |
| 携程&58 | [obproxy的proxyro账号的密码按照集群来设置](https://github.com/oceanbase/oceanbase/issues/878) | 概要设计中 |
| 内部需求 | [ob-operator支持修改 ob 配置](https://github.com/oceanbase/ob-operator/issues/3) | 概要设计中 |
| 内部需求 | [obclient支持更多mysql-test case 的语句](https://github.com/oceanbase/oceanbase/issues/879) | 概要设计中 |
|  |  |  |


## 生态工具研发
| **需求来源** | **                 目标和任务详细描述** | **进展详细描述** |
| --- | --- | --- |
| **六棱镜** | ocp放开备份恢复功能 | 3.3.0-ce, 测试中, 计划5月上旬发布 |
| 广州智通人才 | 支持ob 到ob 的数据同步 | 3.3.0-ce, 测试中, 计划5月上旬发布 |
| 浦发银行 | 支持ob 实时日志同步到kafka上 | 3.3.0-ce, 测试中, 计划5月上旬发布 |
| 浦发银行 | 支持ob 实时日志同步到rocketmq上 | 3.3.0-ce, 测试中, 计划5月上旬发布 |
| 内部需求 | oms 支持json 数据类型 | 3.3.0-ce, 测试中, 计划5月上旬发布 |
| 内部需求 | 优化OCP社区版安装部署 | 3.3.0-ce, 测试中, 计划5月上旬发布, 重写安装脚本, 支持debug模式和更多校验操作. |
| 携程&58 | [独立configure-server ](https://github.com/oceanbase/oceanbase/issues/880) | 编码中 |
| 内部需求 | 优化OMS 社区版安装部署 | 3.3.0-ce, 测试中, 计划5月上旬发布, 支持更多校验操作. |
| 内部需求 | 内部sql 脚本工具开源 | 增强开源可视化定位能力 |

**Comment by longdafeng** [comment]
## 生态合作
### 对接中

- [
](https://mp.weixin.qq.com/s/lGoT1CgyxFUrCapBHTc8IA)
- 2022/4 通过CloudCanal 全量及增量同步 [对接CloudCanal](https://github.com/oceanbase/tutorials-doc/blob/V1.0.0/zh-CN/4.4-0-migrate-data-to-an-oceanbase-database/14.4-13-how-to-use-cloudcanal-to-migrate-and-realtime-sync-data-to-oceanbase.md)
- 2022/5 通过Chunjun(原flinkx) [对接Chunjun](https://github.com/DTStack/chunjun/)
- 2022/5 对接南京基石数据智能运维平台[D-smart大师问诊软件](https://github.com/oceanbase/oceanbase/wiki/www.dfcdata.com.cn)
### 已经完成

- 2022/4 [大数据流水线系统Piflow 完成和OB 对接](https://mp.weixin.qq.com/s/lGoT1CgyxFUrCapBHTc8IA)
- 2022/3 通过Flink CDC 全量及增量同步 [对接Flink CDC](https://github.com/ververica/flink-cdc-connectors/pull/778)
- 2022/3 通过otter 双向实时同步 [对接otter](https://github.com/alibaba/otter)
- 2022/1 [支持K8S operator](https://github.com/oceanbase/ob-operator) , 会持续演进
- 2021/12 [对接SQLancer](https://github.com/sqlancer/sqlancer/commit/6b7c6f62caf10fab4a6195ef6367ea9ebe6f4f45)
- 2021/10 支持实时同步, 会持续演进, [对接canal](https://github.com/oceanbase/canal/releases), [用户手册](https://open.oceanbase.com/docs/tutorials-cn/V1.0.0/10000000000012275)
- 2021/10 [对接Prometheus/Grafana](https://github.com/oceanbase/obagent), [用户手册](https://open.oceanbase.com/docs/observer-cn/V3.1.2/10000000000015598)
- 2021/9 [对接dbeaver](https://github.com/dbeaver/dbeaver/pull/13748)
- 2021/7 [对接DataX](https://github.com/alibaba/datax), [指导手册](https://open.oceanbase.com/docs/tutorials-cn/V1.0.0/10000000000012271)

**Comment by longdafeng** [comment]
# 运营活动
| ![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1650253728768-3d7f44ea-890b-4751-b067-379a678ac876.png#clientId=u3654c908-154c-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=145&id=ua3cbc477&margin=%5Bobject%20Object%5D&name=image.png&originHeight=290&originWidth=510&originalType=binary&ratio=1&rotation=0&showTitle=false&size=188321&status=done&style=none&taskId=u5fa378fe-3912-40b6-aeaa-5aee214da60&title=&width=255) | [我们做了一个分布式数据库](https://open.oceanbase.com/activities/1921895) |
| --- | --- |
| ![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1650253729049-d6de3f23-5444-4355-af04-b83311d370ad.png#clientId=u3654c908-154c-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=142&id=lzzAb&margin=%5Bobject%20Object%5D&name=image.png&originHeight=284&originWidth=502&originalType=binary&ratio=1&rotation=0&showTitle=false&size=219093&status=done&style=none&taskId=u290bb24a-a2be-42e2-9e8a-224ad417615&title=&width=251) | [新数据库时代：DBA发展之路该如何选择](https://open.oceanbase.com/activities) |
| ![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1650253729050-d625c234-bb80-403f-bb63-cf849cd59eac.png#clientId=u3654c908-154c-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=139&id=ufa86a5f2&margin=%5Bobject%20Object%5D&name=image.png&originHeight=278&originWidth=500&originalType=binary&ratio=1&rotation=0&showTitle=false&size=199123&status=done&style=none&taskId=u31cece18-9b10-4b52-afc5-174d79498cc&title=&width=250) | [首期 OceanBase 社区读书会](https://open.oceanbase.com/activities) |
| ![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1650258175554-a621c5cc-459b-47d4-932f-659f5e50ac71.png#clientId=ub2a64bac-a286-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=139&id=ue0ab3b19&margin=%5Bobject%20Object%5D&name=image.png&originHeight=386&originWidth=702&originalType=binary&ratio=1&rotation=0&showTitle=false&size=262064&status=done&style=none&taskId=ub05d0f74-0761-4137-848c-90ff764977d&title=&width=252) | [文档捉虫系列](https://open.oceanbase.com/ask/detail?id=30500054) |
| ![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1650253728835-3376d4db-9d80-45b4-bbaf-5f49eb24ed1d.png#clientId=u3654c908-154c-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=135&id=u23eda856&margin=%5Bobject%20Object%5D&name=image.png&originHeight=270&originWidth=502&originalType=binary&ratio=1&rotation=0&showTitle=false&size=126842&status=done&style=none&taskId=uddd9828b-081e-44c4-bb87-77f543662bc&title=&width=251) | [深入浅出 OceanBase 第二期](https://open.oceanbase.com/activities) |
| ![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1650253730535-116bd7b1-3637-4150-836c-05c0a9e1cb21.png#clientId=u3654c908-154c-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=104&id=u4d2db5c8&margin=%5Bobject%20Object%5D&name=image.png&originHeight=208&originWidth=516&originalType=binary&ratio=1&rotation=0&showTitle=false&size=151408&status=done&style=none&taskId=u48613389-cc8d-4122-b8cd-f837b1b2cab&title=&width=258) | [DBA 老友会](https://open.oceanbase.com/activities) |
| ![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1650253730556-2c730e94-7f4f-444e-9776-53a2a78c63b7.png#clientId=u3654c908-154c-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=141&id=ub2c76c7f&margin=%5Bobject%20Object%5D&name=image.png&originHeight=282&originWidth=516&originalType=binary&ratio=1&rotation=0&showTitle=false&size=206046&status=done&style=none&taskId=ua5c15ffa-9569-40b7-bcab-e8322f5957a&title=&width=258) | [对话 ACE](https://open.oceanbase.com/activities) |
|  |  |
|  |  |

[
](https://open.oceanbase.com/activities/1921895)

**Comment by longdafeng** [comment]
# 精品内容
推荐 
[《开启 JSON 和多模，让生态更多可能 | OceanBase 社区版 3.1.3 发版》](https://open.oceanbase.com/blog/10900260?currentPage=1)
[《首发！OceanBase 改写系列一：OceanBase 查询改写实践概述》](https://open.oceanbase.com/blog/10900257)
[OceanBase 改写系列二: 子查询提升首篇](https://open.oceanbase.com/blog/10900261)

| 开启 JSON 和多模，让生态更多可能 &#124; OceanBase 社区版 3.1.3 发版 |
| --- |
| OceanBase 改写系列二: 子查询提升首篇 |
| 首发！OceanBase 改写系列一：OceanBase 查询改写实践概述 |
| Sysbench 对 OceanBase 开源版3.1.3数据库的 OLTP 性能测试 |
| 首期社区读书会  从《OceanBase数据库系统概念》到3.1.3 社区新版本，我们一起聊聊 OceanBase 那些事 |
| BenchmarkSQL 对 OceanBase 开源版3.1.2性能测试调优探索 |
|  一文详解系列之如何运维 OceanBase 社区版 |
| OceanBase v3.1.3_CE 发版，支持 JSON 数据类型、OBKV HBase 接口 |

**Comment by longdafeng** [comment]
# 4月客户问题精选
| 我用obd部署obagent， 然后从obagent目录中拿到prometheus的配置文件，把内容加到我自己的Prometheus中，然后发现拿不到数据，显示报错 | 这种情况大概率是obagent端口出了问题 可以在集群中查看下observer的端口，有两个端口，一个是默认的，一个是远程连接的，然后对比下文件/home/admin/obagent/conf/config_properties/monagent_pipeline.yaml中monment.ob.sql.port 和 monment.ob.rpc.port 看下是否统一，否则修改一下。建议搭建好监控环境后先检查下端口无误，这样可以避免数据访问不到的问题。 | 2022-04-01 |
| --- | --- | --- |
|  |  | 
|  |  | 
|  |  | 
| 新创建了租户，然后用root用户登录不上，是什么原因呢 | 对于一个租户来讲，不是随便一个客户端就可以连接的，默认情况是本地IP才可以连接，如果想让其他客户端IP也可以连接，需要用变量ob_tcp_invited_nodes设置租户的IP白名单（ALTER TENANT tenantName SET VARIABLES ob_tcp_invited_nodes='%';） | 2022-04-02 |
|  |  | 
|  |  | 
|  |  | 
| 连接obdumper时报错是怎么回事呢？Could not connect to address=(host=10.125.138.175)(port=2883)(type=master):Unknow system variable 'session_track_schema' | 估计你的ob版不是最新的，版本过低就会报错，因为像obdumper和obloader工具只能搭配ob3.12及以上的版本才能正常使用，所以建议直接把ob升级到最高版本，这样会避免发生报错问题。 | 2022-04-06 |
|  |  | 
|  |  | 
| 启动集群的时候报错了：[ERROR] Another app is currently holding the obd lock.Exception ignored in: <bound method MixLock.__del__ of <_lock.MixLock object at 0x2b7d03cd3f98>> NameError: name 'open' is not defined | memory_limit表示数据库内存使用上限，如果memory_limit的值为0，则使用参数memory_limit_percentage的值，否则使用memory_limit的值。该报错说明数据库内存不够用了，而查看配置文件obd-config.yaml发现memory_limit的值是7G,则尝试调大变成8G就可以正常启动了,同时官网上也提到最小为8G才可以，所以建议以后memory_limit的值设置最小为8G. | 2022-04-07 |
|  |  | 
|  |  | 
| 初始化observer报错：[2022-03-23 10:46:54.761115] ERROR [STORAGE] open (ob_local_file_system.cpp:395) [1198][0][Y0-0000000000000000] [lt=2] [dc=0] data file size is too large, (ret=-4184, free_space=44462501888, data_file_size_=53687091200) BACKTRACE:0x97b78ce 0x970a211 0x2228026 0x2227c6b 0x22279e3 0x221bb69 0x80b2b4c 0x80b1c75 0x8683a59 0x7b87931 0x91414b1 0x9137127 0x21e7705 0x2b66de66c555 0x21e63e9 | 从日志可以看出datafile_size或者datafile_disk_percentage设置的大于磁盘剩余的了，所以需要调整下参数，如果是手动部署的ob,参数设置后有可能还有其他报错，比如ERROR [SERVER] init_storage (ob_server.cpp:1692) [9610][0][Y0-0000000000000000] [lt=2] [dc=0] init partition service fail(ret=-4009, storage_env_={data_dir:"/home/admin/oceanbase/store/obdemo", default_block_size:2097152, disk_avail_space:21474836480, datafile_disk_percentage:90, redundancy_level:1, log_spec:{log_dir:"/home/admin/oceanbase/store/obdemo/slog", max_log_size:268435456, log_sync_type:0}, clog_dir:"/home/admin/oceanbase/store/obdemo/clog", ilog_dir:"/home/admin/oceanbase/store/obdemo/ilog", clog_shm_path:"/home/admin/oceanbase/store/obdemo/clog_shm", ilog_shm_path:"/home/admin/oceanbase/store/obdemo/ilog_shm", index_cache_priority:10, user_block_cache_priority:1, user_row_cache_priority:1, fuse_row_cache_priority:1, bf_cache_priority:1, clog_cache_priority:1, index_clog_cache_priority:1, bf_cache_miss_count_threshold:100, ethernet_speed:131072000}) BACKTRACE:0x97b78ce 0x970a211 0x2225fdf 0x2225c2b 0x22259f2 0x7dfadeb 0x91416de 0x9137127 0x21e7705 0x2ba05d0a1555 0x21e63e9 这种情况下很可能是目录出了问题，建议把手动创建的目录结构下清空一下，重新执行一下，一般就可以解决了。所以设置参数大小时不能无限大，还要考虑磁盘空间，如果能用obd部署也会减少报错的可能性。 | 2022-04-08 |
| 类型转换过程中，int转成varchar时报错 Alter non string type not supported | 类型修改过程中有个原则，要转成的类型一定要兼容将要转的类型，比如bigint类型是兼容int类型的，text类型兼容varchar,long blob是兼容blob的，这种情况就可以由下转上，int向上转成bigint, text向上转成varchar, blog向上转成long blob。而像varchar并没有兼容int,也没有兼容char,所以int和char这两种类型都不可以转换成varchar。所以类型转换注意两点：1. 有上下兼容关系。2. 要由下转上。 | 2022-04-11 |
|  |  | 
|  |  | 
| 测试环境部署了三个ob， 同一个zone。查看租户资源的时候只看到一个svr_ip，为什么不是三个svr_ip? | 在一个zone中，同一个资源池中的多个Unit不能分配到同一个server中,而这个zone中部署了三个server,也就是说如果让每个server都分配到unit的话，创建的单个zone下Unit的个数不能少于3个，而一个资源池包含的 unit 个数同时又不能超过单个zone 内 server 的个数，所以该zone中unit的个数只能为3个，只需要把资源池中的UNIT_NUM的值修改为3就可以实现该租户在三个server都占有了资源，这样查看租户资源时就可以看到三个svr_ip了。 | 2022-04-12 |
|  |  | 
|  |  | 
| 当手动启动集群中某个节点（bin/observer）时经常会报错：error while loading shared libraries: libmariadb.so.3: cannot open shared object file: No such file or directory | 这是因为没有将 ob的 LIB 加到环境变量 LD_LIBRARY_PATH 中,所以就类似Java中找不到类，缺少相关jar包，此时需要执行export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:~/oceanbase-ce/lib/ 就可以将ob的 LIB 加到环境变量 LD_LIBRARY_PATH 中。或者执行echo 'export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:~/oceanbase-ce/lib/' >> ~/.bash_profile 写到.bash_profile中都可以解决，需要注意的是路径~/oceanbase-ce/lib/要使用自己安装的ob的实际路径，这个路径并不是固定的。 | 2022-04-13 |
|  |  | 
|  |  | 
| 使用ob的jdbc驱动时，如果使用"jdbcUrl": "&#124;&#124;_dsc_ob10_dsc_&#124;&#124;obdemo:oboracle&#124;&#124;_dsc_ob10_dsc_&#124;&#124;jdbc:oceanbase://127.0.0.1:2883/tpcc?useLocalSessionState=true&allowBatch=true&allowMultiQueries=true&rewriteBatchedStatements=true",经常会出现各种各样的错误 | "jdbcUrl": "&#124;&#124;_dsc_ob10_dsc_&#124;&#124;obdemo:oboracle&#124;&#124;_dsc_ob10_dsc_&#124;&#124;jdbc:oceanbase://127.0.0.1:2883/tpcc?useLocalSessionState=true&allowBatch=true&allowMultiQueries=true&rewriteBatchedStatements=true",这个是对应商业版。而社区版有更简单的写法"jdbcUrl": "jdbc:mysql://127.0.0.1:2883/tpcc?useLocalSessionState=true&allowBatch=true&allowMultiQueries=true&rewriteBatchedStatements=true",也就是说社区版直接用mysql驱动就可以了，这样会避免一些错误。 | 2022-04-14 |
|  |  | 
|  |  | 
|  |  | 
|  |  | 
| load data加载文件时，经常会报错：File not exist | 对于这个错误，初步想到的是，在某个路径下找不到相关文件，然后查看官网文档要求导入文件要放在ob所在机器的目录下，可即使放在了某个目录下，还是不行。后来又发现另一个要求，在load.py文件中hostname要求写某个具体observer的IP,而且port要求写该observer对应的端口，这说明load data导入文件需要直连，而不能通过obproxy访问，所以需要把连接方式修改一下就可以解决了。 | 2022-04-15 |
|  |  | 
|  |  | 
|  |  |

**Comment by yr137157** [comment]
【SOP 系列 01】OBD 部署 OceanBase 集群 yaml 文件修改说明
https://open.oceanbase.com/ask/detail?id=28800052&search=&pageNum=1
【SOP 系列 02】使用 OBD 在线部署 OceanBase 集群
https://open.oceanbase.com/ask/detail?id=28800056&search=&pageNum=1
【SOP 系列 03】使用 OBD 离线部署 OceanBase 集群
https://open.oceanbase.com/ask/detail?id=28800057&search=&pageNum=1
【SOP 系列 04】手动部署 OceanBase 集群
https://open.oceanbase.com/ask/detail?id=28800061&search=&pageNum=1
【SOP 系列 05】报错后如何定位 OceanBase 错误日志
https://open.oceanbase.com/ask/detail?id=28800063&search=&pageNum=1
【SOP 系列 06】手动扩容 OceanBase 集群
https://open.oceanbase.com/ask/detail?id=29700039&search=&pageNum=1
【SOP 系列 07】如何使用 OCP 接管 OBD 部署的 OceanBase 集群
https://open.oceanbase.com/ask/detail?id=30100012&search=&pageNum=1
【SOP 系列 08】分配资源报错时如何准确统计可用（剩余）资源
https://open.oceanbase.com/ask/detail?id=30500006&search=&pageNum=1
